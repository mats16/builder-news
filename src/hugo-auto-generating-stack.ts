import { Stack, StackProps, Duration } from 'aws-cdk-lib';
import * as acm from 'aws-cdk-lib/aws-certificatemanager';
import * as cf from 'aws-cdk-lib/aws-cloudfront';
import { S3Origin } from 'aws-cdk-lib/aws-cloudfront-origins';
import * as codebuild from 'aws-cdk-lib/aws-codebuild';
import * as events from 'aws-cdk-lib/aws-events';
import * as targets from 'aws-cdk-lib/aws-events-targets';
import * as iam from 'aws-cdk-lib/aws-iam';
import * as lambda from 'aws-cdk-lib/aws-lambda';
import { NodejsFunction } from 'aws-cdk-lib/aws-lambda-nodejs';
import * as logs from 'aws-cdk-lib/aws-logs';
import * as s3 from 'aws-cdk-lib/aws-s3';
import * as s3deploy from 'aws-cdk-lib/aws-s3-deployment';
import * as sfn from 'aws-cdk-lib/aws-stepfunctions';
import * as sfnTasks from 'aws-cdk-lib/aws-stepfunctions-tasks';
import { Construct } from 'constructs';

interface HugoStackProps extends StackProps {
  config: {
    cfCname?: string;
    acmArn?: string;
    hugoEnv?: string;
    hugoGoogleAnalytics?: string;
    hugoDisqusShortname?: string;
  };
}

export class HugoStack extends Stack {
  constructor(scope: Construct, id: string, props: HugoStackProps = { config: {} }) {
    super(scope, id, props);

    const { cfCname, acmArn, hugoEnv, hugoGoogleAnalytics, hugoDisqusShortname } = props.config;

    const hugoBucketPath = 'hugo';
    const hugoContentBucketPath = `${hugoBucketPath}/content`;
    const hugoPublicBucketPath = `${hugoBucketPath}/public`;

    const bucket = new s3.Bucket(this, 'Bucket', {
      encryption: s3.BucketEncryption.S3_MANAGED,
      eventBridgeEnabled: true,
    });

    new s3deploy.BucketDeployment(this, 'DeployWebsite', {
      sources: [s3deploy.Source.asset('./hugo')],
      destinationBucket: bucket,
      destinationKeyPrefix: `${hugoBucketPath}/`,
      prune: false,
    });

    const translateStatement = new iam.PolicyStatement({
      actions: ['translate:TranslateText'],
      resources: ['*'],
    });

    const createPostFunction = new NodejsFunction(this, 'CreatePostFunction', {
      description: 'Create new post',
      entry: './src/functions/create-post/index.ts',
      handler: 'handler',
      runtime: lambda.Runtime.NODEJS_14_X,
      architecture: lambda.Architecture.ARM_64,
      timeout: Duration.minutes(3),
      environment: {
        POWERTOOLS_SERVICE_NAME: 'CreatePostFunction',
        POWERTOOLS_METRICS_NAMESPACE: this.stackName,
        POWERTOOLS_TRACER_CAPTURE_RESPONSE: 'false',
        HUGO_CONTENT_BUCKET_NAME: bucket.bucketName,
        HUGO_CONTENT_BUCKET_PATH: hugoContentBucketPath,
      },
      logRetention: logs.RetentionDays.ONE_WEEK,
    });
    bucket.grantReadWrite(createPostFunction, `${hugoContentBucketPath}/*.md`);
    createPostFunction.addToRolePolicy(translateStatement);

    const createThumbnailFunction = new lambda.DockerImageFunction(this, 'CreateThumbnailFunction', {
      description: 'Create thumbnail image and put to S3',
      code: lambda.DockerImageCode.fromImageAsset('./src/functions/create-thumbnail/'),
      architecture: lambda.Architecture.X86_64,
      timeout: Duration.minutes(3),
      environment: {
        POWERTOOLS_SERVICE_NAME: 'CreateThumbnailFunction',
        POWERTOOLS_METRICS_NAMESPACE: this.stackName,
        POWERTOOLS_TRACER_CAPTURE_RESPONSE: 'false',
        HUGO_CONTENT_BUCKET_NAME: bucket.bucketName,
        HUGO_CONTENT_BUCKET_PATH: hugoContentBucketPath,
      },
      logRetention: logs.RetentionDays.ONE_WEEK,
    });
    bucket.grantPut(createThumbnailFunction, `${hugoContentBucketPath}/*.png`);

    const urlRewriteFunction = new cf.Function(this, 'UrlRewriteFunction', {
      code: cf.FunctionCode.fromFile({
        filePath: './src/functions/url-rewrite/index.js',
      }),
    });

    const cfDistribution = new cf.Distribution(this, 'Distribution', {
      comment: 'Builder News',
      domainNames: (typeof cfCname == 'undefined') ? undefined : [cfCname],
      certificate: (typeof acmArn == 'undefined') ? undefined : acm.Certificate.fromCertificateArn(this, 'Certificate', acmArn),
      defaultBehavior: {
        origin: new S3Origin(bucket, { originPath: `/${hugoPublicBucketPath}` }),
        viewerProtocolPolicy: cf.ViewerProtocolPolicy.HTTPS_ONLY,
        functionAssociations: [
          {
            eventType: cf.FunctionEventType.VIEWER_REQUEST,
            function: urlRewriteFunction,
          },
        ],
      },
      defaultRootObject: 'index.html',
      errorResponses: [
        { httpStatus: 403, ttl: Duration.days(1), responsePagePath: '/404.html', responseHttpStatus: 404 },
        { httpStatus: 404, ttl: Duration.days(1), responsePagePath: '/404.html' },
      ],
    });

    const buildProject = new codebuild.Project(this, 'BuildStaticPages', {
      description: 'Hugo - Build static pages',
      source: codebuild.Source.s3({
        bucket: bucket,
        path: `${hugoBucketPath}/`,
      }),
      environment: { buildImage: codebuild.LinuxBuildImage.STANDARD_5_0 },
      timeout: Duration.minutes(10),
      environmentVariables: {
        HUGO_DOWNLOAD_URL: { value: 'https://github.com/gohugoio/hugo/releases/download/v0.97.0/hugo_0.97.0_Linux-64bit.tar.gz' },
        HUGO_BUCKET_NAME: { value: bucket.bucketName },
        HUGO_PUBLIC_BUCKET_PATH: { value: hugoPublicBucketPath },
        HUGO_BASEURL: { value: `https://${cfCname||cfDistribution.distributionDomainName}/` },
        HUGO_PARAMS_ENV: { value: hugoEnv || 'development' },
        HUGO_GOOGLEANALYTICS: { value: hugoGoogleAnalytics || '' },
        HUGO_DISQUSSHORTNAME: { value: hugoDisqusShortname || '' },
      },
      buildSpec: codebuild.BuildSpec.fromObject({
        version: '0.2',
        phases: {
          build: {
            commands: [
              'rm -rf ./public/*',
              'curl -L ${HUGO_DOWNLOAD_URL} | tar zx -C /usr/local/bin',
              'hugo --buildDrafts --buildFuture',
              'aws s3 sync --delete ./public/ s3://${HUGO_BUCKET_NAME}/${HUGO_PUBLIC_BUCKET_PATH}/',
            ],
          },
        },
      }),
    });
    bucket.grantRead(buildProject, `${hugoBucketPath}/*`);
    bucket.grantWrite(buildProject, `${hugoPublicBucketPath}/*`);

    const createEnglishPostTask = new sfnTasks.LambdaInvoke(this, 'Create English Post', {
      lambdaFunction: createPostFunction,
      payload: sfn.TaskInput.fromObject({
        input: sfn.JsonPath.entirePayload,
        lang: 'en',
      }),
    });

    const createJapanesePostTask = new sfnTasks.LambdaInvoke(this, 'Create Japanese Post', {
      lambdaFunction: createPostFunction,
      payload: sfn.TaskInput.fromObject({
        input: sfn.JsonPath.entirePayload,
        lang: 'ja',
      }),
    });

    const createEnglisThumbnailTask = new sfnTasks.LambdaInvoke(this, 'Create Englis Thumbnail', {
      lambdaFunction: createThumbnailFunction,
    });
    createEnglishPostTask.next(createEnglisThumbnailTask);

    const createJapaneseThumbnailTask = new sfnTasks.LambdaInvoke(this, 'Create Japanese Thumbnail', {
      lambdaFunction: createThumbnailFunction,
    });
    createJapanesePostTask.next(createJapaneseThumbnailTask);

    const hugoBuildTask = new sfnTasks.CodeBuildStartBuild(this, 'Hugo Build', {
      integrationPattern: sfn.IntegrationPattern.RUN_JOB,
      project: buildProject,
    });

    const clearCdnCacheTask = new sfnTasks.CallAwsService(this, 'Clear CDN Cache', {
      service: 'CloudFront',
      action: 'createInvalidation',
      parameters: {
        DistributionId: cfDistribution.distributionId,
        InvalidationBatch: {
          'CallerReference.$': '$.SdkResponseMetadata.RequestId',
          'Paths': {
            Items: ['/*'],
            Quantity: 1,
          },
        },
      },
      iamResources: [`arn:aws:cloudfront::${this.account}:distribution/${cfDistribution.distributionId}`],
      iamAction: 'cloudfront:CreateInvalidation',
    });

    const createSummaryTask = new sfn.Parallel(this, 'Create Summary').branch(createJapanesePostTask).branch(createEnglishPostTask);
    createSummaryTask.next(hugoBuildTask).next(clearCdnCacheTask);

    const generateHugoContentsJob = new sfn.StateMachine(this, 'GenerateHugoContents', {
      definition: createSummaryTask,
    });

    new events.Rule(this, 'ScheduledStablePostRule', {
      description: 'Create stable post for Hugo every day',
      schedule: events.Schedule.expression('cron(0 0 ? * MON-SAT *)'),
      targets: [new targets.SfnStateMachine(generateHugoContentsJob, {
        maxEventAge: Duration.hours(1),
        retryAttempts: 3,
        input: events.RuleTargetInput.fromObject({
          time: events.EventField.time,
          isDraft: false,
        }),
      })],
    });

    new events.Rule(this, 'ScheduledDraftPost', {
      description: 'Create draft post for Hugo every day',
      schedule: events.Schedule.expression('cron(0 22 ? * SUN-FRI *)'),
      targets: [new targets.SfnStateMachine(generateHugoContentsJob, {
        maxEventAge: Duration.hours(1),
        retryAttempts: 3,
        input: events.RuleTargetInput.fromObject({
          time: events.EventField.time,
          isDraft: true,
        }),
      })],
    });

    const hugoConfigChanedRule = new events.Rule(this, 'HugoConfigChaned', {
      description: 'Rebuild static pages, because Hugo config changed',
      eventPattern: {
        source: ['aws.s3'],
        detailType: ['Object Created'],
        detail: {
          bucket: {
            name: [bucket.bucketName],
          },
          object: {
            key: [{ prefix: 'hugo/config.' }],
          },
        },
      },
    });
    hugoConfigChanedRule.addTarget(new targets.CodeBuildProject(buildProject));

    this.exportValue(`https://${cfCname||cfDistribution.distributionDomainName}/`, { name: 'Url' });
  }
}
