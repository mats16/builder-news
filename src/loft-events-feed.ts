import * as cdk from 'aws-cdk-lib';
import * as cf from 'aws-cdk-lib/aws-cloudfront';
import * as events from 'aws-cdk-lib/aws-events';
import { LambdaFunction as LambdaFunctionTarget } from 'aws-cdk-lib/aws-events-targets';
import * as lambda from 'aws-cdk-lib/aws-lambda';
import { NodejsFunction } from 'aws-cdk-lib/aws-lambda-nodejs';
import * as logs from 'aws-cdk-lib/aws-logs';
import * as s3 from 'aws-cdk-lib/aws-s3';
import { Construct } from 'constructs';

interface LoftEventsFeedProps {
  bucket: s3.IBucket;
  distribution: cf.IDistribution;
}

export class LoftEventsFeed extends Construct {
  constructor(scope: Construct, id: string, props: LoftEventsFeedProps) {
    super(scope, id);

    const { bucket, distribution } = props;

    const feedKey = 'startup/loft/tokyo/events';

    /** 記事の Markdown を生成する Lambda Function */
    const generateFeedFunction = new NodejsFunction(this, 'GenerateFeedFunction', {
      description: 'Generate feed',
      entry: './src/functions/gen-loft-events-feed/index.ts',
      handler: 'handler',
      runtime: lambda.Runtime.NODEJS_20_X,
      architecture: lambda.Architecture.ARM_64,
      timeout: cdk.Duration.minutes(1),
      environment: {
        BUCKET_NAME: bucket.bucketName,
        FEED_KEY: feedKey,
        DISTRIBUTION_ID: distribution.distributionId,
      },
      logRetention: logs.RetentionDays.ONE_WEEK,
    });

    /** S3 への書き込み権限を付与 */
    bucket.grantReadWrite(generateFeedFunction, feedKey);

    new events.Rule(this, 'ScheduledFeedGeneration', {
      schedule: events.Schedule.rate(cdk.Duration.minutes(10)),
      targets: [new LambdaFunctionTarget(generateFeedFunction, { retryAttempts: 0 })],
    });

  }
}
