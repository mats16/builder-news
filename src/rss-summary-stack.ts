import { Stack, StackProps, Duration } from 'aws-cdk-lib';
import * as cf from 'aws-cdk-lib/aws-cloudfront';
import { S3Origin } from 'aws-cdk-lib/aws-cloudfront-origins';
import * as events from 'aws-cdk-lib/aws-events';
import * as targets from 'aws-cdk-lib/aws-events-targets';
import * as iam from 'aws-cdk-lib/aws-iam';
import * as lambda from 'aws-cdk-lib/aws-lambda';
import { NodejsFunction } from 'aws-cdk-lib/aws-lambda-nodejs';
import * as logs from 'aws-cdk-lib/aws-logs';
import * as s3 from 'aws-cdk-lib/aws-s3';
import { Construct } from 'constructs';

export class RssSummaryStack extends Stack {
  constructor(scope: Construct, id: string, props: StackProps = {}) {
    super(scope, id, props);

    const bucket = new s3.Bucket(this, 'Bucket');

    const translateStatement = new iam.PolicyStatement({
      actions: ['translate:TranslateText'],
      resources: ['*'],
    });

    const createSummaryFunction = new NodejsFunction(this, 'CreateSummaryFunction', {
      description: 'Create summary',
      entry: './src/functions/create-summary/index.ts',
      handler: 'handler',
      runtime: lambda.Runtime.NODEJS_14_X,
      architecture: lambda.Architecture.ARM_64,
      //memorySize: 256,
      timeout: Duration.minutes(5),
      environment: {
        POWERTOOLS_SERVICE_NAME: 'CreateSummaryFunction',
        POWERTOOLS_METRICS_NAMESPACE: this.stackName,
        POWERTOOLS_TRACER_CAPTURE_RESPONSE: 'false',
        RSS_URL: 'https://aws.amazon.com/about-aws/whats-new/recent/feed/',
        BUCKET_NAME: bucket.bucketName,
      },
      logRetention: logs.RetentionDays.TWO_WEEKS,
    });
    bucket.grantPut(createSummaryFunction);
    createSummaryFunction.addToRolePolicy(translateStatement);

    const rule = new events.Rule(this, 'sampleRule', {
      //schedule: events.Schedule.expression('cron(0 23 ? * MON-FRI *)'),
      schedule: events.Schedule.expression('cron(0/5 * ? * * *)'), // test
    });

    rule.addTarget(new targets.LambdaFunction(createSummaryFunction));

    new cf.Distribution(this, 'Distribution', {
      comment: 'Summary',
      defaultBehavior: {
        origin: new S3Origin(bucket),
      },
      enableIpv6: true,
    });

  }
}
