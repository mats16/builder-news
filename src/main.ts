import { App } from 'aws-cdk-lib';
import { RssSummaryStack } from './rss-summary-stack';

// for development, use account/region from cdk cli
const devEnv = {
  account: process.env.CDK_DEFAULT_ACCOUNT,
  region: process.env.CDK_DEFAULT_REGION,
};

const app = new App();

new RssSummaryStack(app, 'rss-summary-dev', { env: devEnv });
// new MyStack(app, 'rss-sumally-prod', { env: prodEnv });

app.synth();