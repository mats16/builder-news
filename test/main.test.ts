import { App } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { RssSummaryStack } from '../src/rss-summary-stack';

test('Snapshot', () => {
  const app = new App();
  const stack = new RssSummaryStack(app, 'test');

  const template = Template.fromStack(stack);
  expect(template.toJSON()).toMatchSnapshot();
});