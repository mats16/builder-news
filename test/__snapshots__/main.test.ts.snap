// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot 1`] = `
Object {
  "Parameters": Object {
    "BootstrapVersion": Object {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": Object {
    "Bucket83908E77": Object {
      "DeletionPolicy": "Retain",
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
    "BucketPolicyE9A3008A": Object {
      "Properties": Object {
        "Bucket": Object {
          "Ref": "Bucket83908E77",
        },
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "s3:GetObject",
              "Effect": "Allow",
              "Principal": Object {
                "CanonicalUser": Object {
                  "Fn::GetAtt": Array [
                    "DistributionOrigin1S3Origin5F5C0696",
                    "S3CanonicalUserId",
                  ],
                },
              },
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    Object {
                      "Fn::GetAtt": Array [
                        "Bucket83908E77",
                        "Arn",
                      ],
                    },
                    "/*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::S3::BucketPolicy",
    },
    "CreateSummaryFunction917C2F11": Object {
      "DependsOn": Array [
        "CreateSummaryFunctionServiceRoleDefaultPolicy4FA02F7C",
        "CreateSummaryFunctionServiceRole12DD7A62",
      ],
      "Properties": Object {
        "Architectures": Array [
          "arm64",
        ],
        "Code": Object {
          "S3Bucket": Object {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "ee0557825be401e935b68611246a9e7f6096e170d5d7eef40258d052d81185ec.zip",
        },
        "Description": "Create summary",
        "Environment": Object {
          "Variables": Object {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
            "BUCKET_NAME": Object {
              "Ref": "Bucket83908E77",
            },
            "POWERTOOLS_METRICS_NAMESPACE": "test",
            "POWERTOOLS_SERVICE_NAME": "CreateSummaryFunction",
            "POWERTOOLS_TRACER_CAPTURE_RESPONSE": "false",
            "RSS_URL": "https://aws.amazon.com/about-aws/whats-new/recent/feed/",
          },
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "CreateSummaryFunctionServiceRole12DD7A62",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
        "Timeout": 300,
      },
      "Type": "AWS::Lambda::Function",
    },
    "CreateSummaryFunctionLogRetention0C6E12B6": Object {
      "Properties": Object {
        "LogGroupName": Object {
          "Fn::Join": Array [
            "",
            Array [
              "/aws/lambda/",
              Object {
                "Ref": "CreateSummaryFunction917C2F11",
              },
            ],
          ],
        },
        "RetentionInDays": 14,
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aFD4BFC8A",
            "Arn",
          ],
        },
      },
      "Type": "Custom::LogRetention",
    },
    "CreateSummaryFunctionServiceRole12DD7A62": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "CreateSummaryFunctionServiceRoleDefaultPolicy4FA02F7C": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "s3:PutObject",
                "s3:PutObjectLegalHold",
                "s3:PutObjectRetention",
                "s3:PutObjectTagging",
                "s3:PutObjectVersionTagging",
                "s3:Abort*",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    Object {
                      "Fn::GetAtt": Array [
                        "Bucket83908E77",
                        "Arn",
                      ],
                    },
                    "/*",
                  ],
                ],
              },
            },
            Object {
              "Action": "translate:TranslateText",
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "CreateSummaryFunctionServiceRoleDefaultPolicy4FA02F7C",
        "Roles": Array [
          Object {
            "Ref": "CreateSummaryFunctionServiceRole12DD7A62",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "Distribution830FAC52": Object {
      "Properties": Object {
        "DistributionConfig": Object {
          "Comment": "Summary",
          "DefaultCacheBehavior": Object {
            "CachePolicyId": "658327ea-f89d-4fab-a63d-7e88639e58f6",
            "Compress": true,
            "TargetOriginId": "testDistributionOrigin1F4407531",
            "ViewerProtocolPolicy": "allow-all",
          },
          "Enabled": true,
          "HttpVersion": "http2",
          "IPV6Enabled": true,
          "Origins": Array [
            Object {
              "DomainName": Object {
                "Fn::GetAtt": Array [
                  "Bucket83908E77",
                  "RegionalDomainName",
                ],
              },
              "Id": "testDistributionOrigin1F4407531",
              "S3OriginConfig": Object {
                "OriginAccessIdentity": Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "origin-access-identity/cloudfront/",
                      Object {
                        "Ref": "DistributionOrigin1S3Origin5F5C0696",
                      },
                    ],
                  ],
                },
              },
            },
          ],
        },
      },
      "Type": "AWS::CloudFront::Distribution",
    },
    "DistributionOrigin1S3Origin5F5C0696": Object {
      "Properties": Object {
        "CloudFrontOriginAccessIdentityConfig": Object {
          "Comment": "Identity for testDistributionOrigin1F4407531",
        },
      },
      "Type": "AWS::CloudFront::CloudFrontOriginAccessIdentity",
    },
    "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aFD4BFC8A": Object {
      "DependsOn": Array [
        "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRoleDefaultPolicyADDA7DEB",
        "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "558dc75fbdc793c8e2f64f85d2d7ed2cb820c727ec7e9b08666ad1af0103992e.zip",
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRoleDefaultPolicyADDA7DEB": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "logs:PutRetentionPolicy",
                "logs:DeleteRetentionPolicy",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRoleDefaultPolicyADDA7DEB",
        "Roles": Array [
          Object {
            "Ref": "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "sampleRule0DFCE8EA": Object {
      "Properties": Object {
        "ScheduleExpression": "cron(0/5 * ? * * *)",
        "State": "ENABLED",
        "Targets": Array [
          Object {
            "Arn": Object {
              "Fn::GetAtt": Array [
                "CreateSummaryFunction917C2F11",
                "Arn",
              ],
            },
            "Id": "Target0",
          },
        ],
      },
      "Type": "AWS::Events::Rule",
    },
    "sampleRuleAllowEventRuletestCreateSummaryFunctionAFFA1EB8C6E57DFA": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "CreateSummaryFunction917C2F11",
            "Arn",
          ],
        },
        "Principal": "events.amazonaws.com",
        "SourceArn": Object {
          "Fn::GetAtt": Array [
            "sampleRule0DFCE8EA",
            "Arn",
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
  },
  "Rules": Object {
    "CheckBootstrapVersion": Object {
      "Assertions": Array [
        Object {
          "Assert": Object {
            "Fn::Not": Array [
              Object {
                "Fn::Contains": Array [
                  Array [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  Object {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
